Add Rust comparison

andrewjradcliffe opened the pull request:
As a result of https://github.com/JuliaLang/julia/issues/47715, I noticed this repository and thought: why not add a Rust comparison?

For this addition to work, one would need to have `rustc` and `cargo` installed.

andrewjradcliffe committed:
Add Rust comparison

As a result of https://github.com/JuliaLang/julia/issues/47715, I
noticed this repository and thought: why not add a Rust comparison?

For this addition to work, one would need to have `rustc` and `cargo`
installed.

LilithHafner commented on src/script_rs/src/main.rs:
```diff
@@ -0,0 +1,64 @@
+    x.resize(n, 0.0);
+    let mut rng = Xoshiro256PlusPlus::from_rng(rand::thread_rng()).unwrap();
+
+    for _ in 0..m {
```
```suggestion
    for _ in 1..m {
```

LilithHafner commented on src/script_rs/src/main.rs:
```diff
@@ -0,0 +1,64 @@
+    x.resize(n, 0.0);
+    let mut rng = Xoshiro256PlusPlus::from_rng(rand::thread_rng()).unwrap();
+
+    for _ in 0..m {
```
```suggestion
    for _ in 1..m {
```
Average over `m` trails, not `m+1` (not that it should matter).

If `0..m-1` is more standard rust style, that's okay too.

LilithHafner replied:
Oops! Just figured out the distinction between `..` and `..=`. This is good as is.

LilithHafner commented on src/script_rs/src/main.rs:
```diff
@@ -0,0 +1,64 @@
+
+    for _ in 0..m {
+        rng.fill(&mut x[..]);
+        y += dotindex(&x);
```
```suggestion
        rng.fill(&mut x[..]);

        y += dotindex(&x);
```
Leave a blank line here for stylistic consistency with other implementations and to make parallels with `f` more clear.

andrewjradcliffe replied:
Updated with `rustfmt`, then added line.

LilithHafner commented on src/script_rs/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "script_rs"
+version = "0.1.0"
+edition = "2021"
```
```suggestion
edition = "2022"
```

andrewjradcliffe replied:
2021 is the latest Rust edition -- see the [Edition book](https://doc.rust-lang.org/stable/edition-guide/editions/index.html), in particular, [Section 1.1](https://doc.rust-lang.org/stable/edition-guide/editions/creating-a-new-project.html)

LilithHafner replied:
I thought this referred to the edition of the package, not of rust.

LilithHafner commented on src/script_rs/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+
+[dependencies]
+rand = "0.8.5"
+rand_xoshiro = "0.6.0"
```
Why did you choose to use `rand_xoshiro` rather than the rust stdlib's random number generation? Performance shouldn't be an issue so long as it is substantially faster than sorting because we subtract the runtime of `g` from the runtime of `f`..

andrewjradcliffe replied:
Rust's stdlib does not include random number generators. `rand` provides a number of things, but not xoshiro generators -- `rand_xoshiro` is part of the rust-random org. The implementations are identical to David Blackman and Sebastiano Vigna's implementations (see linked code, other interesting things at https://prng.di.unimi.it/#intro).

I specifically chose xoshiro256++ as this is the default (since 1.7) generator used by Julia's `Random` stdlib. My objective was to replicate the PRNG behavior as closely as possible between the two. The implementations of Julia's xoshiro256++ and `rand_xoshiro`'s are identical, thus, the statistical properties are identical in that respect; the seeding procedure are slightly different (but both use OS entropy at their core), so that's one source of variation that can't conveniently be locked down. 
Additionally, Julia uses SIMD bulk generation with xoshiro256++ when filling arrays (i.e. for calls of the form `rand!(::AbstractArray{Float64})`) of bits types such as `Float64`, `UInt64`, etc. I am not certain that `rustc` will SIMD the loop body inside the `fill` call -- I'd need to examine the assembly; based on the source code, it does not seem likely unless some very aggressive inlining takes place. Note that it would be straightforward to guarantee if we use `portable_simd` and nightly Rust, but that would still not necessarily yield the same behavior as Julia's bulk generation.

The easy way to force the generated sequences to be identical is to provide identical initial states (256 bits, which Julia takes as 4 `UInt64`s and Rust as a static array of `UInt8`s), and for each `i=1:N, i=1:m` advance the state using the `jump` function. Additionally, to opt out of Julia's SIMD bulk generation, replace the `rand!` calls with something like
```julia
function elementwiserand!(rng, z::AbstractArray{T}) where {T}
    for i ∈ eachindex(z)
        z[i] = rand(rng, T)
    end
    z
end
```

The details above are superfluous for this purpose, but if one is going to benchmark sorting algorithms -- which can be sensitive to patterns (or lack thereof) in data -- it is worthwhile to consider such aspects.

LilithHafner commented on src/script_rs/.gitignore:
```diff
@@ -0,0 +1 @@
+/target
```
I'd rather put this in the top-level .gitignore.

andrewjradcliffe replied:
Done.

LilithHafner reviewed (commented):
Thank you for your contribution! This is great. It ran without hiccup on my machine. Once this merges, I'll regenerate all the data and update the figure.

LilithHafner commented:
Rust is the first non-julia language that is consistently faster than Julia 1.8.3 for large sizes!

See the updated figure at: https://github.com/LilithHafner/InterLanguageSortingComparisons/tree/update-figure

andrewjradcliffe committed:
Reformat with `rustfmt`

andrewjradcliffe committed:
Stylistic consistency

As per https://github.com/LilithHafner/InterLanguageSortingComparisons/pull/3#discussion_r1037890149

andrewjradcliffe committed:
Move to top-level gitignore

andrewjradcliffe commented:
> Rust is the first non-julia language...

Once this PR is merged, I can add a threaded version of the above (very minor modification using `rayon`) -- for when you get around to adding multi-threaded sorting to Julia.
